x-pack/plugins/alerting/server/lib/create_abortable_es_client_factory.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { TransportRequestOptions, TransportResult } from '@elastic/elasticsearch';
+import { SearchResponse } from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import { IScopedClusterClient } from 'src/core/server';
+import type { ESSearchRequest } from 'src/core/types/elasticsearch';
+
+export interface IAbortableEsClient {
+  search: (
+    query: ESSearchRequest,
+    options?: TransportRequestOptions
+  ) => Promise<TransportResult<SearchResponse<unknown>, unknown>>;
+}
+
+export interface IAbortableClusterClient {
+  readonly asInternalUser: IAbortableEsClient;
+  readonly asCurrentUser: IAbortableEsClient;
+}
+export interface CreateAbortableEsClientFactoryOpts {
+  scopedClusterClient: IScopedClusterClient;
+  abortController: AbortController;
+}
+
+export function createAbortableEsClientFactory(opts: CreateAbortableEsClientFactoryOpts) {
+  const { scopedClusterClient, abortController } = opts;
+  return {
+    asInternalUser: {
+      search: async (query: ESSearchRequest, options?: TransportRequestOptions) => {
+        try {
+          const searchOptions = options ?? {};
+          return await scopedClusterClient.asInternalUser.search(query, {
+            ...searchOptions,
+            signal: abortController.signal,
+          });
+        } catch (e) {
+          if (isAbortError(e)) {
+            throw new Error('Search has been aborted due to cancelled execution');
+          }
+          throw e;
+        }
+      },
+    },
+    asCurrentUser: {
+      search: async (query: ESSearchRequest, options?: TransportRequestOptions) => {
+        try {
+          const searchOptions = options ?? {};
+          return await scopedClusterClient.asCurrentUser.search(query, {
+            ...searchOptions,
+            signal: abortController.signal,
+          });
+        } catch (e) {
+          if (isAbortError(e)) {
+            throw new Error('Search has been aborted due to cancelled execution');
+          }
+          throw e;
+        }
+      },
+    },
+  };
+}
+
+function isAbortError(error: Error) {
+  return (
+    error.message === 'Request aborted' ||
```

Comment:
ughhh :-)

Obviously this feels a bit fragile.  Can we check `abortController.signal.aborted` instead?

Reply:
Great idea! Updated in this commit: https://github.com/elastic/kibana/pull/120506/commits/a0f354719863fcb28c4f35eee62729b6e799b399